docs/contributing/coding-standards/js.md

```diff
@@ -5,25 +5,52 @@
 JavaScript files have the same name as the component's folder name. Test files have a `.test` suffix placed before the file extension.
 
 ```
-checkboxes
-├── checkboxes.mjs
-└── checkboxes.test.js
+component
+├── component.mjs
+└── component.test.js
 ```
 
 ## Skeleton
 
 ```js
-import { nodeListForEach } from '../vendor/common.mjs'
+/**
+ * Component name
+ *
+ * @class
+ * @param {HTMLElement | Element} $module - HTML element to use for component
```

Comment:
No not really, from a documentation point of view I hoped it showed we _actually_ want `HTMLElement`. 

Possibly a better example in [**details.mjs**](https://github.com/alphagov/govuk-frontend/blob/fix-types/src/govuk/components/details/details.mjs#L18) where I've put:

```mjs
/**
 * Details component
 *
 * @class
 * @param {HTMLDetailsElement | HTMLElement | Element} $module - HTML element to use for details
 * @this {Details}
 */
```

In an ideal world (modern browser) we'd document only `{HTMLDetailsElement}` but since we feature detect `HTMLDetailsElement`(no polyfill) versus `HTMLElement` I felt it important to show both

Or for for [**skip-link.mjs**](https://github.com/alphagov/govuk-frontend/blob/fix-types/src/govuk/components/skip-link/skip-link.mjs#L9) we need a way of saying it's only for `HTMLAnchorElement`

```mjs
/**
 * Skip link component
 *
 * @class
 * @param {HTMLAnchorElement | Element} $module - HTML element to use for skip link
 * @this {SkipLink}
 */
```

Would your preference be to keep it as `Element` everywhere?

Reply:
@36degrees I've given it a go in https://github.com/alphagov/govuk-frontend/pull/2987/commits/8f7093e3245f920e6ae6960a52233f1c105db2ea 😊 

There are a few snags but far fewer than I thought

### Before hitting `$input.checked`
The compiler wants to check what type it is

```mjs
if (!($input instanceof HTMLInputElement)) {
  return
}
```

### Before hitting `$input.checked`, `$input.form` + `$input.name`
The compiler wants to check for inputs

```mjs
if (!($input instanceof HTMLInputElement)) {
  return
}
```

### Before allowing `click` and `keydown` listeners
The compiler wants elements compatible with mouse/keyboard events

```mjs
if (!($tab instanceof HTMLAnchorElement)) {
  return
}
```

Reply:
Hmm, that's not quite what I had in mind – it's only where we had `HTMLElement | Element` that I expected would change, as one is a subclass of the other it feels redundant to include both.

I'd still expect us to use the more specific types wherever we can.

Reply:
Hey that's absolutely fine, I'll get it tweaked

What are your thoughts on adding the `Array.from()` polyfill to replace `nodeListForEach()`?

I might not be able to run `tsc --build` in some places if JSDoc receives `Element` but subsequent code calls a function that only accepts `{HTMLAnchorElement}` if no **instanceof** checks have taken place

Reply:
I'm not against it – it's consistent with what we do everywhere else – but not sure if we can make that change without it being breaking? 🤔 

Was `tsc --build` working without the changes in 95b298e?

Reply:
Ah is adding a new polyfill a breaking change?

It'll work with `"strictFunctionTypes": false` but losing some JSDoc type checks (see https://github.com/alphagov/govuk-frontend/pull/2971#pullrequestreview-1208120890), otherwise it thinks you've made a mistake if you go on to call methods only found on `HTMLElement`

This is because `nodeListForEach()` accepts `Element` (not `HTMLElement`) via **querySelectorAll()**

I'll add back stricter types where I can and push the commit back up 👍 

<img width="826" alt="HTMLElement when Element type is param" src="https://user-images.githubusercontent.com/415517/206156518-e130275c-1a6b-4420-9e14-90ef1925f3c9.png">

Reply:
Here's another example

Receiving an `{Element}` and passing it straight into a function that only accept `{HTMLElement}`

But since we only need `.querySelector()` and `.classList` I can keep `{Element}` on **.setExpanded()**

<img width="1058" alt="Problem with using HTMLElement on setExpanded method" src="https://user-images.githubusercontent.com/415517/206157868-2924acaa-eef6-483d-a383-28ac4d3c7b04.png">

Reply:
> Ah is adding a new polyfill a breaking change?

No, but removing `nodeListForEach` _might_ be. Having said that, we could keep it around in case anyone is importing it from `common`, and just not use it ourselves?

Reply:
Yeah that's fair

Means doing a quick `Array.from()` on the **NodeList** and we can use regular for loops

Have a look now in https://github.com/alphagov/govuk-frontend/pull/2987/commits/8ae699d33136cdba3b08b69104836d823cf44d8a?

Reply:
@romaricpascal Sadly due to `{Element}` in **nodeListForEach()** that stops `HTMLInputElement` as you'd suggested in this comment: https://github.com/alphagov/govuk-frontend/commit/95b298ed5e0ed457ce0bac3eea00a0a48850c622#r92272894 (old commit)

Might be another one that `Array.from()` will solve. See it in action via this PR branch:

* https://github.com/alphagov/govuk-frontend/pull/2971

Should work in any IDE or editor that supports `tsconfig.json` (WebStorm, VS Code etc)

Reply:
I may have a way out that doesn't require us to bring in a polyfill for `Array.from` and risk breaking 🎉 This requires us to type `nodeListForEach` using [TypeScript Generics](https://www.typescriptlang.org/docs/handbook/2/generics.html) so the trail of types can be followed. 

Opened [a PR with changes and explanations](https://github.com/alphagov/govuk-frontend/pull/3096) which we could cherry pick on here 😄

Reply:
Thanks again for this Romaric, was brilliant

We'll see `Element` as the default still but should be able to avoid lots of `instanceof` checks where more specific types can be inferred now instead 🙌